src/mml_utils.py

```diff
@@ -0,0 +1,129 @@
+from bayeslite import bql_quote_name
+from bdbcontrib.population_method import population_method
+from enum import Enum
+from string import Template
+
+import json
+import jsonschema
+import pkgutil
+
+MML_SCHEMA = json.loads(
+    pkgutil.get_data('bdbcontrib', 'mml.schema.json'))
+
+
+class StatType(Enum):
+    CATEGORICAL = 1
+    NUMERICAL = 2
+    IGNORE = 3
+
+
+@population_method(population_to_bdb=0, population_name=1)
+def guess_types(bdb, table):
+    """Guesses stattypes of a given table.
+
+    Returns a dictionary from column names to the guessed stattype.
+    """
+    types = {}
+    for col in _column_names(bdb, table):
+        cursor = bdb.sql_execute(
+            'SELECT DISTINCT %s FROM %s'
+            % (bql_quote_name(col), bql_quote_name(table)))
+        # TODO(asilvers): We don't necessarily need to read all of these in,
+        # and for some datasets this may be prohibitive. But it's fine for now.
+        vals = {row[0] for row in cursor if row[0] not in [None, '']}
+        types[col] = _type_given_vals(vals)
+    return types
+
+
+def _column_names(bdb, table):
+    """Returns the column names of a table"""
+    with bdb.savepoint():
+        cursor = bdb.sql_execute('PRAGMA table_info(%s)'
+                                 % (bql_quote_name(table),))
+        return list(row[1] for row in cursor)
+
+
+def _type_given_vals(vals):
+    """Returns a guess as to the stattype of a column given its values.
+
+    These heuristics are utterly unprincipled and simply seem to work well
+    on a minimal set of test datasets. They should be treated fluidly and
+    improved as failures crop up.
+    """
```

Comment:
TL;DR: The end goal was to get to something like https://github.com/asilversempirical/bdbcontrib/pull/3 which would be much trickier while leaning on bayeslite.guess.

The other half of what I was hoping to do was provide feedback on what guess was thinking. That made me a bit nervous when considering using bayesdb_guess_stattypes for a couple reasons. For one I knew the interface would have to change to provide that feedback and I didn't relish updating all the callers, nor was it obvious you would want to modify that API.

I also think there needs to be a fundamental difference between this guess and bayeslite.guess. As it stands, guess(*) tries to do things without user intervention. That necessitates more complexity and ad-hocness. That's a fine goal, but right now it's not working for people -- if it gets a single column wrong there's no solid way to incrementally improve, and it's not at the point yet where it consistently nails it.

Whereas the aim of this utility is more of a jumping off point for writing a schema. And for that I think it's better to have a slightly worse guess implementation that fits in a user's head.